make.js

```diff
@@ -193,331 +167,370 @@ function parseDataFile(filename) {
-            prerequisiteApis[api.name].push(api.calls.find(elem => elem.name === "GetGroup"));
-            prerequisiteApis[api.name].push(api.calls.find(elem => elem.name === "ApplyToGroup"));
-            prerequisiteApis[api.name].push(api.calls.find(elem => elem.name === "InviteToGroup"));
-            prerequisiteApis[api.name].push(api.calls.find(elem => elem.name === "BlockEntity"));
+function getOrCreateFeatureGroup(name) {
+    if (!SDKFeatureGroups.hasOwnProperty(name)) {
+        SDKFeatureGroups[name] = {
+            "name": name,
+            "datatypes": {},
+            "calls": [],
+            "sortedClasses": [], // subset of 'datatypes' that are classes (datatype.isclass = true), sorted such that type dependencies are honored
+            "dictionaryEntryTypes": {} // subset of 'datatypes' that appear as values in a map (property.collection = map)
         }
     }
+    return SDKFeatureGroups[name];
 }
 
-function getEnumTypes(datatypes) {
-    var enumtypes = [];
+function populateSDKFeatureGroups(apis) {
 
-    for (var typeIdx in datatypes) // Add all types and their dependencies
-        if (datatypes[typeIdx].isenum)
-            enumtypes.push(datatypes[typeIdx]);
-    return enumtypes;
-}
+    // Create special "Shared" feature group manually. It will hold datatypes which are referenced in multiple SDKFeatureGroups.
+    getOrCreateFeatureGroup("Shared");
 
-function getSortedClasses(api) {
-    var sortedClasses = [];
-    var addedTypes = new Set();
-
-    var addType = function (datatype) {
-        if (addedTypes.has(datatype.name) || datatype.isenum)
-            return;
-        // In C++, dependent types must be defined first
-        if (datatype.properties) {
-            for (var propIdx = 0; propIdx < datatype.properties.length; propIdx++) {
-                var property = datatype.properties[propIdx];
-                if ((property.isclass || property.isenum) && api.datatypes.hasOwnProperty(property.actualtype)) {
-                    addType(api.datatypes[property.actualtype]);
-                }
-            }
-        }
-        addedTypes.add(datatype.name);
-        sortedClasses.push(datatype);
-    };
+    for (var apiIndex = 0; apiIndex < apis.length; apiIndex++) {
+        var api = apis[apiIndex];
+        var apiDatatypes = api["datatypes"];
 
-    // Add all types and their dependencies
-    for (var typeIdx in api.datatypes) {
-        addType(api.datatypes[typeIdx]);
-    }
+        for (var callIndex = 0; callIndex < api.calls.length; ++callIndex) {
 
-    // Annotate classes
-    annotateClasses(sortedClasses, api);
+            var call = api.calls[callIndex];
 
-    return sortedClasses;
-}
+            // Mangle name of "Classic" API calls to avoid conflicts. Note that this is in addition to the global & featureGroup prefixes that will
+            // be added to the public SDK APIs.
+            // TODO should we only do this when there is a conflict? or for all classic APIs?

```

Comment:
>// TODO should we only do this when there is a conflict? or for all classic APIs? [](http://example.com/codeflow?start=12&length=81)

@jasonsandlin  thoughts on this?  Should we always add Client/Server/Admin to the class API call names? or only when there are duplicates to avoid conflict?

Reply:
its better to add client/server/admin always so callers know its a classic API.  Otherwise its the caller can't easily tell if its an entity or classic.  the caller needs to know at times since its a different backend data set.  it just makes it more clear what we're up to by adding subgroups

Reply:
Makes sense, will always include that then